modules/extensions/src/terrain/terrain-cover.ts

```diff
@@ -0,0 +1,251 @@
+    this.tile = getTile(owner);
+  }
+
+  get id() {
+    return this.owner.id;
+  }
+
+  get isActive(): boolean {
+    const owner = this.owner.getCurrentLayer();
+    return owner !== null && !owner.lifecycle.startsWith('Finalized');
+  }
+
+  shouldUpdate({
+    owner,
+    viewport,
+    layers,
+    layerNeedsRedraw
+  }: {
+    owner?: Layer;
+    viewport?: Viewport;
+    layers?: Layer[];
+    layerNeedsRedraw?: Record<string, boolean>;
+  }): boolean {
+    if (owner) {
+      this.owner = owner;
+    }
+    const sizeChanged = viewport ? this._updateViewport(viewport) : false;
+
+    let layersChanged = layers ? this._updateLayers(layers) : false;
+
+    if (layerNeedsRedraw) {
+      for (const id of this.layers) {
+        if (layerNeedsRedraw[id]) {
+          layersChanged = true;
+          // console.log('layer needs redraw', id);
+          break;
+        }
+      }
+    }
+
+    return layersChanged || sizeChanged;
+  }
+
+  /** Compare layers with the last version. Only rerender if necessary. */
+  private _updateLayers(layers: Layer[]): boolean {
+    let needsRedraw = false;
+    layers = getIntersectingLayers(this.tile, layers);
+
+    if (layers.length !== this.layers.length) {
+      needsRedraw = true;
+      // console.log('layers count changed', this.layers.length, '>>', layers.length);
+    } else {
+      for (let i = 0; i < layers.length; i++) {
+        const id = layers[i].id;
+        if (id !== this.layers[i]) {
+          needsRedraw = true;
+          // console.log('layer added/removed', id);
+          break;
+        }
+      }
+    }
+    if (needsRedraw) {
+      this.layers = layers.map(layer => layer.id);
+    }
+    return needsRedraw;
+  }
+
+  /** Compare viewport and terrain bounds with the last version. Only rerender if necesary. */
+  private _updateViewport(viewport: Viewport): boolean {
+    const owner = this.owner;
+    let needsRedraw = false;
+    const newBounds = owner.getBounds();
+    if (
+      // The terrain layer's bounds has changed
+      this.bounds !== newBounds ||
+      // If the terrain layer is not bound to a tile, increase texture size based on zoom
+      (!this.tile && Math.abs(viewport.zoom - this.zoom) >= 1)
+    ) {
+      needsRedraw = true;
+      // console.log('bounds changed', this.bounds, '>>', newBounds);
+      this.bounds = newBounds;
+      // console.log('zoom changed', this.zoom, '>>', Math.ceil(viewport.zoom));
+      this.zoom = Math.ceil(viewport.zoom);
+    }
+
+    if (newBounds) {
+      const leftBottomCommon = owner.projectPosition(newBounds[0], {viewport});
+      const topRightCommon = owner.projectPosition(newBounds[1], {viewport});
+      this.commonBounds = [leftBottomCommon, topRightCommon];
+    } else {
+      this.commonBounds = null;
+    }
+
+    if (needsRedraw) {
+      this.renderViewport = getRenderViewport(owner, this.zoom, viewport.isGeospatial);
+    }
+    return needsRedraw;
+  }
+
+  get renderTexture(): Framebuffer | null {
```

Comment:
- Do accessors provide a substantial advantage over getters here? I generally make myself think twice before using getters, to avoid too much spooky action.
- These are called Texture but return Framebuffers is that helpful?